deltachat-contact-tools/src/lib.rs

```diff
@@ -29,10 +29,116 @@ use std::fmt;
 use std::ops::Deref;
 
 use anyhow::bail;
+use anyhow::format_err;
+use anyhow::Context as _;
 use anyhow::Result;
+use chrono::DateTime;
 use once_cell::sync::Lazy;
 use regex::Regex;
 
+// TODOs to clean up:
+// - Check if sanitizing is done correctly everywhere
+// - Apply lints everywhere (https://doc.rust-lang.org/cargo/reference/workspaces.html#the-lints-table)
+
+#[derive(Debug)]
+/// A Contact, as represented in a VCard.
+pub struct VcardContact {
+    /// The email address, vcard property `email`
+    pub addr: String,
+    /// The contact's display name, vcard property `fn`
+    pub display_name: String,
+    /// The contact's public PGP key, vcard property `key`
+    pub key: Option<String>,
+    /// The contact's profile photo (=avatar), vcard property `photo`
+    pub profile_photo: Option<String>,
+    /// The timestamp when the vcard was created / last updated, vcard property `rev`
+    pub timestamp: Result<u64>,
+}
+
+pub fn parse_vcard(vcard: String) -> Result<Vec<VcardContact>> {
+    fn remove_prefix<'a>(s: &'a str, prefix: &str) -> Option<&'a str> {
+        let start_of_s = s.get(..prefix.len())?;
+
+        if start_of_s.eq_ignore_ascii_case(prefix) {
+            s.get(prefix.len()..)
+        } else {
+            None
+        }
+    }
+    fn vcard_property<'a>(s: &'a str, property: &str) -> Option<&'a str> {
+        let remainder = remove_prefix(s, property)?;
+
+        // TODO this doesn't handle the case where there are quotes around a colon
+        let (_params, value) = remainder.split_once(':')?;
+        Some(value)
+    }
+    fn parse_datetime(datetime: Option<&str>) -> Result<u64> {
+        let datetime = datetime.context("No timestamp in vcard")?;
+
+        // According to https://www.rfc-editor.org/rfc/rfc6350#section-4.3.5, the timestamp
+        // is in ISO.8601.2004 format. DateTime::parse_from_rfc3339() apparently parses
+        // ISO.8601, but fails to parse any of the examples given.
+        // So, instead just parse using a format string.
+        let datetime =
+            DateTime::parse_from_str(datetime, "%Y%m%dT%H%M%S%#z") // Parses 19961022T140000Z, 19961022T140000-05, or 19961022T140000-0500
+                .or_else(|_| DateTime::parse_from_str(datetime, "%Y%m%dT%H%M%S"))?; // Parses 19961022T140000
+        let timestamp = datetime.timestamp().try_into()?;
```

Comment:
Probably can be simplified to `datetime.timestamp().try_into()`

Reply:
No, because then no error conversion is done (try_into() gives us some custom error but we need an anyhow error)